extensions/kafka-eight-simpleConsumer/src/main/java/io/druid/firehose/kafka/KafkaSimpleConsumer.java

```diff
@@ -0,0 +1,406 @@
+
+    return response != null ? filterAndDecode(response.messageSet(topic, partitionId), offset) : EMPTY_MSGS;
+  }
+
+  private void stopConsumer()
+  {
+    if (consumer != null) {
+      try {
+        consumer.close();
+        log.info("stop consumer[%s][%s], leaderBroker[%s]", topic, partitionId, leaderBroker);
+      }
+      catch (Exception e) {
+        log.warn(e, "stop consumer[%s][%s] failed", topic, partitionId);
+      }
+      finally {
+        consumer = null;
+      }
+    }
+  }
+
+  // stop the consumer
+  public void stop()
+  {
+    stopConsumer();
+    log.info("KafkaSimpleConsumer[%s][%s] stopped", topic, partitionId);
+  }
+
+  private PartitionMetadata findLeader() throws InterruptedException
+  {
+    for (KafkaBroker broker : replicaBrokers) {
+      SimpleConsumer consumer = null;
+      try {
+        log.info("Finding new leader from Kafka brokers, try broker %s:%s", broker.host, broker.port);
+        consumer = new SimpleConsumer(broker.host, broker.port, soTimeout, bufferSize, leaderLookupClientId);
+        TopicMetadataResponse resp = consumer.send(new TopicMetadataRequest(Collections.singletonList(topic)));
+
+        List<TopicMetadata> metaData = resp.topicsMetadata();
+        for (TopicMetadata item : metaData) {
+          if (topic.equals(item.topic())) {
+            for (PartitionMetadata part : item.partitionsMetadata()) {
+              if (part.partitionId() == partitionId) {
+                return part;
+              }
+            }
+          }
+        }
+      }
+      catch (Exception e) {
+        ensureNotInterrupted(e);
+        log.warn(
+            e, "error communicating with Kafka Broker [%s] to find leader for [%s] - [%s]", broker, topic, partitionId
+        );
+      }
+      finally {
+        if (consumer != null) {
+          try {
+            consumer.close();
+          }
+          catch (Exception e) {
+          }
+        }
+      }
+    }
+
+    return null;
+  }
+
+  private Broker findNewLeader(Broker oldLeader) throws InterruptedException
+  {
+    long retryCnt = 0;
+    while (true) {
+      PartitionMetadata metadata = findLeader();
+      if (metadata != null) {
+        replicaBrokers.clear();
+        for (Broker replica : metadata.replicas()) {
+          replicaBrokers.add(new KafkaBroker(replica.host(), replica.port()));
+        }
+
+        log.debug("Got new Kafka leader metadata : [%s], previous leader : [%s]", metadata, oldLeader);
+        Broker newLeader = metadata.leader();
+        if (newLeader != null) {
+          // We check the retryCnt here as well to make sure that we have slept a little bit
+          // if we don't notice a change in leadership.  Note, that this means that this method
+          // is hopefully only being called when we believe that leadership has changed.  Otherwise
+          // there is a big chunk of sleep being added for no reason.
+          if (oldLeader == null || !isCurrLeader(newLeader) || retryCnt != 0) {
```

Comment:
Thanks for pointing that out, the original wording is quite confusing. So I renamed the method `isCurrLeader` to `isValidNewLeader`. So the logic here is: we shall return the newLeader, if 1) on first retry, found newLeader different to oldLeader, 2) on >=1 retry, found a not null newLeader. This logic is mainly to avoid assigning to oldLeader on first retry, since Zookeeper might haven't yet realized the oldLeader going down.